From 90b6cc4ae649b60e314157de81204821bb539243 Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Thu, 4 May 2023 15:52:54 +0200
Subject: [PATCH 6/9] Add support for user-data ntp

This needs to happen early so time is right for https later
---
 lib/tiny-cloud/init  | 40 ++++++++++++++++++++++++++++++++++++++++
 tests/init-main.test | 43 ++++++++++++++++++++++++++++++++++++++++++-
 2 files changed, 82 insertions(+), 1 deletion(-)

diff --git a/lib/tiny-cloud/init b/lib/tiny-cloud/init
index 9596889..0ac88f8 100644
--- a/lib/tiny-cloud/init
+++ b/lib/tiny-cloud/init
@@ -19,6 +19,7 @@ INIT_ACTIONS_EARLY="
 INIT_ACTIONS_MAIN="
     save_userdata
     set_hostname
+    userdata_ntp
     set_ssh_keys
 "
 INIT_ACTIONS_FINAL="
@@ -150,6 +151,35 @@ init__save_userdata() {
     rm "$tmpfile"
 }
 
+init__userdata_ntp() {
+    local ntp_enabled="$(get_userdata ntp enabled)"
+    if [ "$ntp_enabled" != "yes" ] && [ "$ntp_enabled" != "true" ]; then
+        return
+    fi
+    local ntp_client="$(get_userdata ntp ntp_client)"
+    local svc= pkg=
+    case "$ntp_client" in
+        busybox)
+            svc=ntpd
+            ;;
+        chrony|"")
+            pkg=chrony
+            svc=chronyd
+            ;;
+        openntpd)
+            pkg=openntpd
+            svc=openntpd
+            ;;
+    esac
+    if [ -n "$pkg" ]; then
+        $MOCK apk add "$pkg"
+    fi
+    if [ -n "$svc" ]; then
+        $MOCK rc-update add "$svc" default
+        $MOCK rc-service "$svc" start
+    fi
+}
+
 ### init-final functions
 
 init__run_userdata() {
@@ -191,5 +221,15 @@ userdata_type() {
     fi
 }
 
+get_userdata() {
+    local userdata="$TINY_CLOUD_VAR/user-data"
+    if ! [ -f "$userdata" ]; then
+        init__save_userdata
+    fi
+    if [ "$(userdata_type)" = "script" ]; then
+        return
+    fi
+    yx -f "$userdata" "$@" 2>/dev/null
+}
 # ...load user-data type-specific init functions / vars
 # TODO
diff --git a/tests/init-main.test b/tests/init-main.test
index 0cfe770..05cda06 100755
--- a/tests/init-main.test
+++ b/tests/init-main.test
@@ -10,7 +10,10 @@ init_tests \
 	set_hostname \
 	set_ssh_keys \
 	save_userdata_plain \
-	save_userdata_compressed
+	save_userdata_compressed \
+	userdata_ntp \
+	userdata_ntp_busybox \
+	userdata_ntp_openntpd
 
 
 set_nocloud_meta() {
@@ -87,3 +90,41 @@ save_userdata_compressed_body() {
 		fi
 	done
 }
+
+userdata_ntp_body() {
+	fake_userdata_nocloud <<-EOF
+		ntp:
+		  enabled: true
+	EOF
+	CLOUD=nocloud atf_check \
+		-o match:"apk add.*chrony" \
+		-o match:"rc-update .*chronyd" \
+		-o match:"rc-service .*chronyd" \
+		sh -c ". \"$lib\"; init__userdata_ntp"
+}
+
+userdata_ntp_busybox_body() {
+	fake_userdata_nocloud <<-EOF
+		ntp:
+		  enabled: true
+		  ntp_client: busybox
+	EOF
+	CLOUD=nocloud atf_check \
+		-o not-match:"apk add" \
+		-o match:"rc-update .*ntpd" \
+		-o match:"rc-service .*ntpd" \
+		sh -c ". \"$lib\"; init__userdata_ntp"
+}
+
+userdata_ntp_openntpd_body() {
+	fake_userdata_nocloud <<-EOF
+		ntp:
+		  enabled: true
+		  ntp_client: openntpd
+	EOF
+	CLOUD=nocloud atf_check \
+		-o match:"apk add.*openntpd" \
+		-o match:"rc-update .*openntpd" \
+		-o match:"rc-service .*openntpd" \
+		sh -c ". \"$lib\"; init__userdata_ntp"
+}
-- 
2.40.1

